### PR TITLE
SAA-668 updating the manage attendnaces job to check for attendance records that need locking down.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
This covers the work for the overnight JOB which manages attendances, their expiry period and whether they are LOCKED to further edits in the activities service.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt [hunk 1]
```diff
@@ -47,7 +47,7 @@ data class Attendance(
   var recordedBy: String? = null,
 
   @Enumerated(EnumType.STRING)
-  var status: AttendanceStatus = AttendanceStatus.WAITING,
+  private var status: AttendanceStatus = AttendanceStatus.WAITING,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt:50]: This should not be directly mutable.  We should be calling methods/functions on the entity to transition the status states.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/ManageAttendanceRecordsJobIntegrationTest.kt [hunk 2]
```diff
@@ -95,7 +97,7 @@ class ManageAttendanceRecordsJobIntegrationTest : IntegrationTestBase() {
 
   @Sql("classpath:test_data/seed-activity-id-5.sql")
   @Test
-  fun `No attendance records are created for gym induction AM`() {
+  fun `No attendance records are created for gym induction AM when attendance not required on activity`() {
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/ManageAttendanceRecordsJobIntegrationTest.kt:100]: Added a bit more context to the test description.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AttendancesServiceTest.kt [hunk 3]
```diff
@@ -161,16 +155,16 @@ class AttendancesServiceTest {
...
-    assertThat(attendance.status).isEqualTo(AttendanceStatus.COMPLETED)
+    assertThat(attendance.status()).isEqualTo(AttendanceStatus.COMPLETED)
     with(attendance.attendanceReason!!) {
       assertThat(code).isEqualTo(AttendanceReasonEnum.OTHER)
     }
     assertThat(attendance.otherAbsenceReason).isEqualTo("other absence reason")
   }
 
   @Test
-  fun `other absence reason is not set when attendance reason does not equal "OTHER"`() {
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AttendancesServiceTest.kt:172]: Double quotes in the test description was throwing a compilation warning.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt [hunk 4]
```diff
@@ -76,8 +83,51 @@ class ManageAttendancesService(
...
+            when {
+              dateOfAttendance.between(yesterday.minusWeeks(2), yesterday) -> {
+                unlockedAttendance.takeIf { it.status(AttendanceStatus.COMPLETED) }
+              }
+
+              else -> unlockedAttendance
+            }?.let {
+              runCatching {
+                attendanceRepository.saveAndFlush(unlockedAttendance.lock())
+              }.onFailure {
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt:116]: Are we going to run into the same issues I have had? i.e. This job is updating attendance to LOCKED, and will emit the same sync event as attendance-updated. Maybe that is ok? Mike will probably throw these away - as the only change will be the paylock flag in NOMIS attendance, and that should already have been set.

**Review by timharrison-moj — APPROVED** [review]
LGTM. Tests look to confirm the correct behaviour - but I wasn't sure from looking at the job code.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt (hunk 4 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt:116]: The only entities we are touching here are attendances so we should only ever get the attendance updated event.  I'll double check this though.

**Pull request merged**